File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
from collections import deque
import sys

read = sys.stdin.readline

V = int(read())
N = [[] for _ in range(V + 1)]
visited = [False] * (V + 1)
distance = [0] * (V + 1)

for _ in range(V):
    nodes = list(map(int, read().split()))
    index = 0
    s = nodes[index]
    index += 1
    while True:
        e = nodes[index]
        if e == -1:
            break
        w = nodes[index + 1]
        N[s].append((e, w))
        index += 2

def BFS(num):
    q = deque()
    q.append(num)
    visited[num] = True

    while q:
        p = q.popleft()
        for i in N[p]:
            if not visited[i[0]]:
                q.append(i[0])
                visited[i[0]] = True
                distance[i[0]] = distance[p] + i[1]

BFS(1)
max = 1
for i in range(2, V + 1):
    if distance[i] > distance[max]:
        max = i

visited = [False] * (V + 1)
distance = [0] * (V + 1)
BFS(max)
distance.sort(reverse = True)
print(distance[0])

File: 백준/Gold/1167. 트리의 지름/트리의 지름.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main{
	static boolean[] visited;
	static ArrayList<Node>[] A;
	static int[] distance;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer stk;
		int v = Integer.parseInt(br.readLine());
		
		A = new ArrayList[v + 1];
		for(int i = 1; i <= v; i++)
			A[i] = new ArrayList<Node>();
			
		for(int i = 0; i < v; i++) {
			stk = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(stk.nextToken());
			
			while(true) {
				int e = Integer.parseInt(stk.nextToken());
				if(e == - 1)
					break;
				int w = Integer.parseInt(stk.nextToken());
				A[s].add(new Node(e, w));
			}
		}
		
		visited = new boolean[v + 1];
		distance = new int[v + 1];
		
		BFS(1);
		int max = 1;
		
		for(int i = 2; i <= v; i++)
		{
			if(distance[max] < distance[i])
				max = i;
		}
		
		visited = new boolean[v + 1];
		distance = new int[v + 1];
		BFS(max);
		
		Arrays.sort(distance);
		System.out.println(distance[v]);
	}
	
	public static void BFS(int num) {
		Queue<Integer> q = new LinkedList<Integer>();
		q.add(num);
		visited[num] = true;
		
		while(!q.isEmpty())
		{
			int now = q.poll();
			for(Node i : A[now])
			{
				if(!visited[i.end]) {
					visited[i.end] = true;
					q.add(i.end);
					distance[i.end] = distance[now] + i.weight;
				}
			}
		}
	}
}

class Node{
	int end;
	int weight;
	
	Node(int end, int weight){
		this.end = end;
		this.weight = weight;
	}
}